<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>TOPMed Analysis Workshop</title>
  <meta name="description" content="Course materials for the TOPMed analysis workshop">
  <meta name="generator" content="bookdown 0.4 and GitBook 2.6.7">

  <meta property="og:title" content="TOPMed Analysis Workshop" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for the TOPMed analysis workshop" />
  <meta name="github-repo" content="UW-GAC/topmed_workshop_2017" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="TOPMed Analysis Workshop" />
  
  <meta name="twitter:description" content="Course materials for the TOPMed analysis workshop" />
  



<meta name="date" content="2017-08-08">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="analysis-pipeline.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>3</b> Computing a GRM</a></li>
<li class="chapter" data-level="4" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>4</b> PC-Relate</a><ul>
<li class="chapter" data-level="4.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>4.1</b> KING</a></li>
<li class="chapter" data-level="4.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>4.2</b> PC-AiR</a></li>
<li class="chapter" data-level="4.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>4.3</b> PC-Relate</a></li>
<li class="chapter" data-level="4.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>4.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="4.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise"><i class="fa fa-check"></i><b>4.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html"><i class="fa fa-check"></i><b>5</b> Phenotype Harmonization</a><ul>
<li class="chapter" data-level="5.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#import-phentoype-files-into-r"><i class="fa fa-check"></i><b>5.1</b> Import phentoype files into R</a></li>
<li class="chapter" data-level="5.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#run-some-quick-diagnostics-on-the-files"><i class="fa fa-check"></i><b>5.2</b> Run some quick diagnostics on the files</a></li>
<li class="chapter" data-level="5.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#use-using-mixed-models-to-compare-studies"><i class="fa fa-check"></i><b>5.3</b> Use using mixed models to compare studies</a><ul>
<li class="chapter" data-level="5.3.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#create-an-annotated-data-frame"><i class="fa fa-check"></i><b>5.3.1</b> Create an Annotated Data Frame</a></li>
<li class="chapter" data-level="5.3.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#obtain-the-genetic-relatedness-matrix"><i class="fa fa-check"></i><b>5.3.2</b> Obtain the genetic relatedness matrix</a></li>
<li class="chapter" data-level="5.3.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-mixed-model-without-harmonization-unit"><i class="fa fa-check"></i><b>5.3.3</b> Fit a mixed model without harmonization unit</a></li>
<li class="chapter" data-level="5.3.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-model-with-harmonization-unit"><i class="fa fa-check"></i><b>5.3.4</b> Fit a model with harmonization unit</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#final-considerations"><i class="fa fa-check"></i><b>5.4</b> Final considerations</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><i class="fa fa-check"></i><b>6</b> Finding unharmonized TOPMed study phenotypes on dbGaP</a><ul>
<li class="chapter" data-level="6.1" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#ways-to-get-topmed-phenotype-data"><i class="fa fa-check"></i><b>6.1</b> Ways to get TOPMed phenotype data</a></li>
<li class="chapter" data-level="6.2" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#dbgap-accession-lingo"><i class="fa fa-check"></i><b>6.2</b> dbGaP accession lingo</a></li>
<li class="chapter" data-level="6.3" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#dbgap-advanced-search-tools"><i class="fa fa-check"></i><b>6.3</b> dbGaP advanced search tools</a><ul>
<li class="chapter" data-level="6.3.1" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#entrez-advanced-search"><i class="fa fa-check"></i><b>6.3.1</b> <a href="https://www.ncbi.nlm.nih.gov/gap/advanced">Entrez advanced search</a></a></li>
<li class="chapter" data-level="6.3.2" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#faceted-advanced-search"><i class="fa fa-check"></i><b>6.3.2</b> <a href="https://www.ncbi.nlm.nih.gov/projects/gapsolr/facets.html">Faceted advanced search</a></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>7</b> Association tests</a><ul>
<li class="chapter" data-level="7.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>7.1</b> Null model</a></li>
<li class="chapter" data-level="7.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>7.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="7.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>7.3</b> Exercises</a></li>
<li class="chapter" data-level="7.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>7.4</b> Sliding window tests</a><ul>
<li class="chapter" data-level="7.4.1" data-path="association-tests.html"><a href="association-tests.html#exercise-1"><i class="fa fa-check"></i><b>7.4.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="association-tests.html"><a href="association-tests.html#annotation-based-aggregate-tests"><i class="fa fa-check"></i><b>7.5</b> Annotation-based aggregate tests</a><ul>
<li class="chapter" data-level="7.5.1" data-path="association-tests.html"><a href="association-tests.html#working-with-variant-annotation"><i class="fa fa-check"></i><b>7.5.1</b> Working with variant annotation</a></li>
<li class="chapter" data-level="7.5.2" data-path="association-tests.html"><a href="association-tests.html#defining-gene-ranges-for-aggregation"><i class="fa fa-check"></i><b>7.5.2</b> Defining “gene” ranges for aggregation</a></li>
<li class="chapter" data-level="7.5.3" data-path="association-tests.html"><a href="association-tests.html#aggregating-topmed-variants-into-genic-units"><i class="fa fa-check"></i><b>7.5.3</b> Aggregating TOPMed variants into genic units</a></li>
<li class="chapter" data-level="7.5.4" data-path="association-tests.html"><a href="association-tests.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>7.5.4</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="7.5.5" data-path="association-tests.html"><a href="association-tests.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>7.5.5</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="7.5.6" data-path="association-tests.html"><a href="association-tests.html#exercise-2"><i class="fa fa-check"></i><b>7.5.6</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html"><i class="fa fa-check"></i><b>8</b> Analysis Pipeline</a><ul>
<li class="chapter" data-level="8.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-a-local-cluster"><i class="fa fa-check"></i><b>8.1</b> Running on a local cluster</a></li>
<li class="chapter" data-level="8.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-aws-batch"><i class="fa fa-check"></i><b>8.2</b> Running on AWS Batch</a><ul>
<li class="chapter" data-level="8.2.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#log-into-aws-docker-image"><i class="fa fa-check"></i><b>8.2.1</b> Log into AWS docker image</a></li>
<li class="chapter" data-level="8.2.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#cd-to-working-directory-and-create-config-file"><i class="fa fa-check"></i><b>8.2.2</b> cd to working directory and create config file</a></li>
<li class="chapter" data-level="8.2.3" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#print-out-aws-commands-if-executing-the-pipeline"><i class="fa fa-check"></i><b>8.2.3</b> Print out AWS commands if executing the pipeline</a></li>
<li class="chapter" data-level="8.2.4" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-pipeline"><i class="fa fa-check"></i><b>8.2.4</b> Execute the pipeline</a></li>
<li class="chapter" data-level="8.2.5" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#monitor-the-jobs"><i class="fa fa-check"></i><b>8.2.5</b> Monitor the jobs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="analysis-commons.html"><a href="analysis-commons.html"><i class="fa fa-check"></i><b>9</b> Analysis Commons</a><ul>
<li class="chapter" data-level="9.1" data-path="analysis-commons.html"><a href="analysis-commons.html#outline"><i class="fa fa-check"></i><b>9.1</b> Outline</a></li>
<li class="chapter" data-level="9.2" data-path="analysis-commons.html"><a href="analysis-commons.html#web-interface-and-running-an-analysis-application"><i class="fa fa-check"></i><b>9.2</b> Web Interface and Running an Analysis Application</a><ul>
<li class="chapter" data-level="9.2.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-1-run-a-single-variant-analysis."><i class="fa fa-check"></i><b>9.2.1</b> Exercise 1) Run a single variant analysis.</a></li>
<li class="chapter" data-level="9.2.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5."><i class="fa fa-check"></i><b>9.2.2</b> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="analysis-commons.html"><a href="analysis-commons.html#command-line-interface"><i class="fa fa-check"></i><b>9.3</b> Command line interface</a><ul>
<li class="chapter" data-level="9.3.1" data-path="analysis-commons.html"><a href="analysis-commons.html#log-in-to-awi"><i class="fa fa-check"></i><b>9.3.1</b> Log in to AWI</a></li>
<li class="chapter" data-level="9.3.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-3-navigate-directories-make-output-directory-examine-files"><i class="fa fa-check"></i><b>9.3.2</b> Exercise 3) Navigate directories, make output directory, examine files</a></li>
<li class="chapter" data-level="9.3.3" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-4-run-single-variant-analysis-from-command-line-using-bash-script"><i class="fa fa-check"></i><b>9.3.3</b> Exercise 4) Run single variant analysis from command line using bash script</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="analysis-commons.html"><a href="analysis-commons.html#writing-your-own-apps"><i class="fa fa-check"></i><b>9.4</b> Writing your own Apps</a><ul>
<li class="chapter" data-level="9.4.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform"><i class="fa fa-check"></i><b>9.4.1</b> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</a></li>
<li class="chapter" data-level="9.4.2" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-6-make-qq-plot"><i class="fa fa-check"></i><b>9.4.2</b> Optional Exercise 6) Make QQ plot</a></li>
<li class="chapter" data-level="9.4.3" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-7-run-conditional-analysis"><i class="fa fa-check"></i><b>9.4.3</b> Optional Exercise 7) Run conditional analysis</a></li>
<li class="chapter" data-level="9.4.4" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set"><i class="fa fa-check"></i><b>9.4.4</b> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">TOPMed Analysis Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="analysis-commons" class="section level1">
<h1><span class="header-section-number">9</span> Analysis Commons</h1>
<div id="outline" class="section level2">
<h2><span class="header-section-number">9.1</span> Outline</h2>
<ul>
<li>Introduction to web-interface</li>
<li>Running a single variant analysis</li>
<li>Workflows and monitoring jobs</li>
<li>Running aggregate tests (SKAT)</li>
<li>Run batch jobs from the command line</li>
<li>Writing your own Apps</li>
</ul>
</div>
<div id="web-interface-and-running-an-analysis-application" class="section level2">
<h2><span class="header-section-number">9.2</span> Web Interface and Running an Analysis Application</h2>
<div id="exercise-1-run-a-single-variant-analysis." class="section level3">
<h3><span class="header-section-number">9.2.1</span> Exercise 1) Run a single variant analysis.</h3>
<p>Log into <a href="http://dnanexus.com" class="uri">http://dnanexus.com</a> using the user name and password listed on the handout.<br />
Should be in the form of Username:<strong>topmed_#</strong> and Password:<strong>Topmed_#</strong>. <em>Ignore warning about default billing account.</em> Navigate to and select <strong>(dcc:tools/genesis_v0.7)</strong></p>
<p>File inputs:<br />
* phenofile -&gt; phenotype/1KG_pheno.csv<br />
* genotypefile -&gt; genotypes/1KG_phase3_subset_chr1.gds<br />
* kinship -&gt; kinship/1KG_kins.Rda<br />
* Note: orange aggregation, annotation and genefile can be left empty</p>
<p>Parameter inputs:<br />
* output folder: output/YOURFOLDERNAME<br />
* outcome <em>(Column name of the outcome variable)</em>: outcome<br />
* covariates <em>(case sepecific)</em>: Population,sex<br />
* prefix for output filename: single_chr1<br />
* test_type: Single<br />
* pheno_id: sample.id<br />
* Note: Other options can be left as their defaults, some are only used for aggreagate tests * Note: The job may finish instantaneously if you don’t change the output file name. It knows that you are running the exact same job and will just reuse results from previous analyses.</p>
</div>
<div id="exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5." class="section level3">
<h3><span class="header-section-number">9.2.2</span> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</h3>
<p><em>Italic</em> inputs below are the same as single variant; update the parameters &amp; files to change to a SKAT test. Go to the monitor tab. Click on the Name of a job ( or someone’s ) that successfully completed the single variant analysis, then click “Launch as new Job” and modify the inputs.</p>
<p>File inputs:<br />
* <em>phenofile -&gt; phenotype/1KG_pheno.csv</em><br />
* <em>genotypefile -&gt; genotypes/1KG_phase3_subset_chr1.gds</em><br />
* <em>kinship -&gt; kinship/1KG_kins.Rda</em><br />
* annotation -&gt; annotation/1KG_annotation_CHR1.txt<br />
* genefile -&gt; aggregation/AggUnit_CHR1_ucscgene.csv</p>
<p>Parameter inputs:<br />
* <em>outcome: outcome</em><br />
* <em>covariates: Population,sex</em><br />
* <em>pheno_id: sample.id</em><br />
* output folder: output/YOURFOLDERNAME<br />
* outputfilename: skat_chr1_geneBased_CADDgt2<br />
* test_type: SKAT<br />
* snp_filter: CADD_phred&gt;2<br />
* min_mac:0<br />
* top_maf: 0.05<br />
* weights: c(1,25)</p>
</div>
</div>
<div id="command-line-interface" class="section level2">
<h2><span class="header-section-number">9.3</span> Command line interface</h2>
<p>References:<br />
* Command Line Interface <a href="https://wiki.dnanexus.com/Command-Line-Client/Quickstart">Quickstart</a><br />
* Index of <a href="https://wiki.dnanexus.com/Command-Line-Client/Index%20of%20dx%20Commands">dx commands</a></p>
<div id="log-in-to-awi" class="section level3">
<h3><span class="header-section-number">9.3.1</span> Log in to AWI</h3>
<p><strong>Replace topmed_## with the user ID from your handout</strong></p>
<pre><code>$ ssh topmed_##@34.212.243.167
You will be prompted for your password, e.g. Topmed_## (Note capitalization)
_Please ignore login warnings

$ source /usr/local/dx-toolkit/environment</code></pre>
<pre><code>$ dx login  --timeout 2h
    Enter the following at the prompts
        username: topmed_##
        password: Topmed_##
        project:dcc ( type 0 to select dcc )

You can select or change project once you are logged in
$ dx select dcc</code></pre>
</div>
<div id="exercise-3-navigate-directories-make-output-directory-examine-files" class="section level3">
<h3><span class="header-section-number">9.3.2</span> Exercise 3) Navigate directories, make output directory, examine files</h3>
<ul>
<li>File paths: &lt;project&gt;:/path/to/file.txt</li>
<li>Example: dcc:/phenotypes/1KG_pheno.csv</li>
</ul>
<p>List directory contents:</p>
<pre><code>$ dx select dcc
$ dx ls
$ dx ls /tools
$ dx ls dcc:/tools</code></pre>
<p>Get results from project</p>
<pre><code>$ dx download dcc:/phenotype/1KG_pheno.csv
$ ls
$ head 1KG_pheno.csv</code></pre>
</div>
<div id="exercise-4-run-single-variant-analysis-from-command-line-using-bash-script" class="section level3">
<h3><span class="header-section-number">9.3.3</span> Exercise 4) Run single variant analysis from command line using bash script</h3>
<p>Open the single_multichrom.sh bash script and edit to replace the output directory “YOURNAME” to your folder</p>
<pre><code>$ dx describe tools/genesis_v0.7</code></pre>
<p>Either edit using nano</p>
<pre><code>$ nano single_multichrom.sh 

Run the App.  Will loop over 2 chromosomes running the single variant analyses
$ ./single_multichrom.sh</code></pre>
</div>
</div>
<div id="writing-your-own-apps" class="section level2">
<h2><span class="header-section-number">9.4</span> Writing your own Apps</h2>
<div id="exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform" class="section level3">
<h3><span class="header-section-number">9.4.1</span> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</h3>
<p>Use app wizard to create template</p>
<pre><code>$ dx-app-wizard

App Name: make_residuals
Title []: Create inverse normal transformed residuals

1st input name (&lt;ENTER&gt; to finish): phenofile
Label (optional human-readable name) []: CSV phenotype file
Choose a class (&lt;TAB&gt; twice for choices): file
This is an optional parameter [y/n]: n

2nd input name (&lt;ENTER&gt; to finish): model
Label (optional human-readable name) []: model for creating residuals (e.g. outcome~age+Population )
Choose a class (&lt;TAB&gt; twice for choices): string
This is an optional parameter [y/n]: n

3rd input name (&lt;ENTER&gt; to finish): prefix
Label (optional human-readable name) []: Output filename prefix
Choose a class (&lt;TAB&gt; twice for choices): string
This is an optional parameter [y/n]: n

4th input name (&lt;ENTER&gt; to finish): &lt;ENTER&gt;

1st output name (&lt;ENTER&gt; to finish): output
Label (optional human-readable name) []: 
Choose a class (&lt;TAB&gt; twice for choices): file

Timeout policy [48h]: 1h
Programming language: bash

*Use defaults for other options*</code></pre>
<p>Look at the files created by the wizard</p>
<pre><code>cd make_residuals/
ls
more dxapp.json </code></pre>
<p>Edit App executable to run an R script</p>
<pre><code>$ vi src/make_residuals.sh
main() {

    echo &quot;Value of phenofile: &#39;$phenofile&#39;&quot;
    echo &quot;Value of model: &#39;$model&#39;&quot;
    echo &quot;Value of prefix: &#39;$prefix&#39;&quot;

    dx download &quot;$phenofile&quot; -o phenofile

    Rscript /make_resid.R $model

    output=$(dx upload output --brief)

    dx-jobutil-add-output output &quot;$output&quot; --class=file
    dx mv ${output} ${prefix}.csv
}
</code></pre>
<p>Create an R script that does the ‘work’ <code>$ vi resources/make_resid.R</code></p>
<pre><code>args&lt;-commandArgs(TRUE)
model &lt;- as.formula(args[1])
print(model)
pheno = read.csv(&quot;phenofile&quot;,as.is=T)
pheno$resid = residuals(lm(model,data=pheno))
pheno$invnt_resid =  with(pheno,qnorm((rank(resid,na.last=&quot;keep&quot;)-0.5)/sum(!is.na(resid))))

write.csv(pheno,file=&quot;output&quot;,row.names=F)</code></pre>
<p>Build the App</p>
<pre><code>$ dx build -f make_residuals --destination=output/YOURNAME/make_residuals</code></pre>
<p>Run the App</p>
<pre><code>$ dx run output/YOURNAME/make_residuals -iphenofile=phenotype/1KG_pheno.csv \
-imodel=outcome~sex+Population -iprefix=1kg_pheno_invnt \
--destination=output/YOURNAME --yes</code></pre>
<p>Monitor Progress</p>
<pre><code>$ dx watch jobid</code></pre>
</div>
<div id="optional-exercise-6-make-qq-plot" class="section level3">
<h3><span class="header-section-number">9.4.2</span> Optional Exercise 6) Make QQ plot</h3>
<p>Make QQ plot of your single variant results.<br />
Select results from the multiple chromosome run (chr21 and chr22).</p>
<p>You will need to identify the p-value column name. To view the results file try these options:</p>
<ol style="list-style-type: decimal">
<li><p><a href="https://wiki.dnanexus.com/Command-Line-Client/Index-of-dx-Commands#download">dx download</a> to download the results for viewing.</p></li>
<li><p>View file through web interface using Visualize ( next to Monitor near top of the page ) and select <a href="https://platform.dnanexus.com/projects/F5jVpJ80JXGQV51P8GqVxPPQ/visualize#"><em>Gzipped File Previewer</em></a></p></li>
<li><p>Pipe zipped file though regular linux commands <a href="https://wiki.dnanexus.com/Command-Line-Client/Index-of-dx-Commands#head--">dx cat</a> to view column names</p></li>
</ol>
<pre><code>$ dx cat output/folder/file | gunzip | head</code></pre>
<p>Once you know the name of the p-value column, run qqplot first through web interface and then try running interactivly from the web interface then from the command line.</p>
<pre><code>$ dx run tools/qqplot</code></pre>
<p><em>Note: the plot label must not contain spaces.</em></p>
</div>
<div id="optional-exercise-7-run-conditional-analysis" class="section level3">
<h3><span class="header-section-number">9.4.3</span> Optional Exercise 7) Run conditional analysis</h3>
<p>Find the name of one associated variant in the single snp results and rerun the single variant analysis conditioning on that variant (e.g. 22:17105517).<br />
<em>Note that the output file name cannot contain a colon (e.g. output file name cannot be single_chr22_single_22:17105517, try single_chr22_single_22_17105517 instead).</em></p>
</div>
<div id="optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set" class="section level3">
<h3><span class="header-section-number">9.4.4</span> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</h3>
<p>This process requires two steps, one to extract the LD for all variants in the region and one to create the plot. Sequencing data sets often contain variants not in external refernce panels, so it is helpful to create your own LD reference.</p>
<p>Step 1: Run GILD (<strong>G</strong>DS <strong>I</strong>nto <strong>LD</strong>) App (tools/gild_v1)</p>
<p>File inputs: * gds_file -&gt; genotypes/1KG_phase3_subset_chr22.gds</p>
<p>Parameter inputs:</p>
<ul>
<li>lead_snp -&gt; 22:17105517</li>
<li>start_pos -&gt; 1</li>
<li>stop_pos -&gt; 51237069</li>
<li>label for results file -&gt; “LD_chr22” <em>output_LD_filename</em></li>
<li>output/YOURNAME</li>
</ul>
<p><em>Note: this can take 10-15 mins to complete</em></p>
<p>Step 2: Run AssocPlot (tools/assocplot)</p>
<p>File inputs:</p>
<ul>
<li>datafile -&gt; single variant association results output for chr22</li>
<li>ldfile -&gt; Output file from Step 1 with .ld suffix</li>
</ul>
<p>Parameter inputs (Minimum required to have the App run successfully with GENESIS output):</p>
<ul>
<li>Output folder -&gt; output/YOURNAME</li>
<li>Marker Column Name -&gt; snpID</li>
<li>P value Column Name -&gt; Score.pval</li>
<li>Index SNP -&gt; 22:17105517</li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="analysis-pipeline.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
