<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>SISG Module 12: Computational Pipeline for WGS Data</title>
  <meta name="description" content="Course materials for SISG Module 12">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="SISG Module 12: Computational Pipeline for WGS Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for SISG Module 12" />
  <meta name="github-repo" content="UW-GAC/topmed_workshop_2018" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="SISG Module 12: Computational Pipeline for WGS Data" />
  
  <meta name="twitter:description" content="Course materials for SISG Module 12" />
  



<meta name="date" content="2018-07-20">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="computing-a-grm.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i><b>1.1</b> Schedule</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#resources"><i class="fa fa-check"></i><b>1.2</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>3</b> Computing a GRM</a></li>
<li class="chapter" data-level="4" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>4</b> PC-Relate</a><ul>
<li class="chapter" data-level="4.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>4.1</b> KING</a></li>
<li class="chapter" data-level="4.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>4.2</b> PC-AiR</a></li>
<li class="chapter" data-level="4.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>4.3</b> PC-Relate</a></li>
<li class="chapter" data-level="4.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>4.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="4.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise"><i class="fa fa-check"></i><b>4.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html"><i class="fa fa-check"></i><b>5</b> Phenotype Harmonization</a><ul>
<li class="chapter" data-level="5.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#inspect-individual-study-data-in-r"><i class="fa fa-check"></i><b>5.1</b> Inspect individual study data in R</a></li>
<li class="chapter" data-level="5.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#compare-study-values"><i class="fa fa-check"></i><b>5.2</b> Compare study values</a></li>
<li class="chapter" data-level="5.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#using-mixed-models-to-compare-studies"><i class="fa fa-check"></i><b>5.3</b> Using mixed models to compare studies</a><ul>
<li class="chapter" data-level="5.3.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#create-an-annotated-data-frame"><i class="fa fa-check"></i><b>5.3.1</b> Create an Annotated Data Frame</a></li>
<li class="chapter" data-level="5.3.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#obtain-the-genetic-relatedness-matrix"><i class="fa fa-check"></i><b>5.3.2</b> Obtain the genetic relatedness matrix</a></li>
<li class="chapter" data-level="5.3.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-mixed-model-without-study"><i class="fa fa-check"></i><b>5.3.3</b> Fit a mixed model without study</a></li>
<li class="chapter" data-level="5.3.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-model-with-study"><i class="fa fa-check"></i><b>5.3.4</b> Fit a model with study</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#final-considerations"><i class="fa fa-check"></i><b>5.4</b> Final considerations</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>6</b> Association tests</a><ul>
<li class="chapter" data-level="6.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>6.1</b> Null model</a></li>
<li class="chapter" data-level="6.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>6.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="6.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>6.3</b> Exercises</a></li>
<li class="chapter" data-level="6.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>6.4</b> Sliding window tests</a><ul>
<li class="chapter" data-level="6.4.1" data-path="association-tests.html"><a href="association-tests.html#exercise-1"><i class="fa fa-check"></i><b>6.4.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="association-tests.html"><a href="association-tests.html#annotation-based-aggregate-tests"><i class="fa fa-check"></i><b>6.5</b> Annotation-based aggregate tests</a><ul>
<li class="chapter" data-level="6.5.1" data-path="association-tests.html"><a href="association-tests.html#working-with-variant-annotation"><i class="fa fa-check"></i><b>6.5.1</b> Working with variant annotation</a></li>
<li class="chapter" data-level="6.5.2" data-path="association-tests.html"><a href="association-tests.html#aggregating-and-filtering-variants-using-annotation"><i class="fa fa-check"></i><b>6.5.2</b> Aggregating and filtering variants using annotation</a></li>
<li class="chapter" data-level="6.5.3" data-path="association-tests.html"><a href="association-tests.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>6.5.3</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="6.5.4" data-path="association-tests.html"><a href="association-tests.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>6.5.4</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="6.5.5" data-path="association-tests.html"><a href="association-tests.html#exercise-2"><i class="fa fa-check"></i><b>6.5.5</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html"><i class="fa fa-check"></i><b>7</b> Analysis Pipeline</a><ul>
<li class="chapter" data-level="7.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-a-local-cluster"><i class="fa fa-check"></i><b>7.1</b> Running on a local cluster</a></li>
<li class="chapter" data-level="7.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-aws-batch"><i class="fa fa-check"></i><b>7.2</b> Running on AWS Batch</a><ul>
<li class="chapter" data-level="7.2.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#log-into-aws-docker-image"><i class="fa fa-check"></i><b>7.2.1</b> Log into AWS docker image</a></li>
<li class="chapter" data-level="7.2.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#download-the-docker-helper-functions"><i class="fa fa-check"></i><b>7.2.2</b> Download the docker helper functions</a></li>
<li class="chapter" data-level="7.2.3" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#cd-to-working-directory-and-create-config-file"><i class="fa fa-check"></i><b>7.2.3</b> cd to working directory and create config file</a></li>
<li class="chapter" data-level="7.2.4" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-python-helper-function-to-run-the-docker-image"><i class="fa fa-check"></i><b>7.2.4</b> Execute the python helper function to run the docker image</a></li>
<li class="chapter" data-level="7.2.5" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#print-out-aws-commands-if-executing-the-pipeline"><i class="fa fa-check"></i><b>7.2.5</b> Print out AWS commands if executing the pipeline</a></li>
<li class="chapter" data-level="7.2.6" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-pipeline"><i class="fa fa-check"></i><b>7.2.6</b> Execute the pipeline</a></li>
<li class="chapter" data-level="7.2.7" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#monitor-the-jobs"><i class="fa fa-check"></i><b>7.2.7</b> Monitor the jobs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="using-docker.html"><a href="using-docker.html"><i class="fa fa-check"></i><b>8</b> Using Docker</a><ul>
<li class="chapter" data-level="8.1" data-path="using-docker.html"><a href="using-docker.html#prerequisites"><i class="fa fa-check"></i><b>8.1</b> Prerequisites</a></li>
<li class="chapter" data-level="8.2" data-path="using-docker.html"><a href="using-docker.html#general-steps"><i class="fa fa-check"></i><b>8.2</b> General Steps</a><ul>
<li class="chapter" data-level="8.2.1" data-path="using-docker.html"><a href="using-docker.html#download-the-docker-image"><i class="fa fa-check"></i><b>8.2.1</b> Download the Docker Image</a></li>
<li class="chapter" data-level="8.2.2" data-path="using-docker.html"><a href="using-docker.html#create-working-directory"><i class="fa fa-check"></i><b>8.2.2</b> Create Working Directory</a></li>
<li class="chapter" data-level="8.2.3" data-path="using-docker.html"><a href="using-docker.html#download-the-docker-helper-functions-1"><i class="fa fa-check"></i><b>8.2.3</b> Download the docker helper functions</a></li>
<li class="chapter" data-level="8.2.4" data-path="using-docker.html"><a href="using-docker.html#running-rstudio"><i class="fa fa-check"></i><b>8.2.4</b> Running RStudio</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="using-docker.html"><a href="using-docker.html#examples"><i class="fa fa-check"></i><b>8.3</b> Examples</a><ul>
<li class="chapter" data-level="8.3.1" data-path="using-docker.html"><a href="using-docker.html#example-1"><i class="fa fa-check"></i><b>8.3.1</b> Example 1</a></li>
<li class="chapter" data-level="8.3.2" data-path="using-docker.html"><a href="using-docker.html#example-2"><i class="fa fa-check"></i><b>8.3.2</b> Example 2</a></li>
<li class="chapter" data-level="8.3.3" data-path="using-docker.html"><a href="using-docker.html#example-3"><i class="fa fa-check"></i><b>8.3.3</b> Example 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="analysis-commons.html"><a href="analysis-commons.html"><i class="fa fa-check"></i><b>9</b> Analysis Commons</a><ul>
<li class="chapter" data-level="9.1" data-path="analysis-commons.html"><a href="analysis-commons.html#outline"><i class="fa fa-check"></i><b>9.1</b> Outline</a></li>
<li class="chapter" data-level="9.2" data-path="analysis-commons.html"><a href="analysis-commons.html#web-interface-and-running-an-analysis-application"><i class="fa fa-check"></i><b>9.2</b> Web Interface and Running an Analysis Application</a><ul>
<li class="chapter" data-level="9.2.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-1-run-a-single-variant-analysis."><i class="fa fa-check"></i><b>9.2.1</b> Exercise 1) Run a single variant analysis.</a></li>
<li class="chapter" data-level="9.2.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5."><i class="fa fa-check"></i><b>9.2.2</b> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="analysis-commons.html"><a href="analysis-commons.html#command-line-interface"><i class="fa fa-check"></i><b>9.3</b> Command line interface</a><ul>
<li class="chapter" data-level="9.3.1" data-path="analysis-commons.html"><a href="analysis-commons.html#log-in-to-awi"><i class="fa fa-check"></i><b>9.3.1</b> Log in to AWI</a></li>
<li class="chapter" data-level="9.3.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-3-navigate-directories-make-output-directory-examine-files"><i class="fa fa-check"></i><b>9.3.2</b> Exercise 3) Navigate directories, make output directory, examine files</a></li>
<li class="chapter" data-level="9.3.3" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-4-run-single-variant-analysis-from-command-line-using-bash-script"><i class="fa fa-check"></i><b>9.3.3</b> Exercise 4) Run single variant analysis from command line using bash script</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="analysis-commons.html"><a href="analysis-commons.html#writing-your-own-apps"><i class="fa fa-check"></i><b>9.4</b> Writing your own Apps</a><ul>
<li class="chapter" data-level="9.4.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform"><i class="fa fa-check"></i><b>9.4.1</b> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</a></li>
<li class="chapter" data-level="9.4.2" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-6-make-qq-plot"><i class="fa fa-check"></i><b>9.4.2</b> Optional Exercise 6) Make QQ plot</a></li>
<li class="chapter" data-level="9.4.3" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set"><i class="fa fa-check"></i><b>9.4.3</b> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SISG Module 12: Computational Pipeline for WGS Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gds-format" class="section level1">
<h1><span class="header-section-number">2</span> GDS format</h1>
<p>GDS is Genomic Data Structure, a storage format that can efficiently store genomic data and provide fast random access to subsets of the data. For more information on GDS for sequence data, read the <a href="https://github.com/zhengxwen/SeqArray/blob/master/vignettes/SeqArrayTutorial.Rmd">SeqArray package vignette</a>.</p>
<div id="exploring-a-gds-file" class="section level2">
<h2><span class="header-section-number">2.1</span> Exploring a GDS file</h2>
<p>To use the R packages developed at the DCC for sequence data, we first need to convert a VCF file to GDS. (If the file is BCF, use <a href="bcftools">https://samtools.github.io/bcftools/bcftools.html</a> to convert to VCF.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqArray)
data.path &lt;-<span class="st"> &quot;https://github.com/UW-GAC/analysis_pipeline/raw/master/testdata&quot;</span>
vcffile &lt;-<span class="st"> &quot;data/1KG_phase3_subset_chr1.vcf.gz&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(<span class="st">&quot;data&quot;</span>)) <span class="kw">dir.create</span>(<span class="st">&quot;data&quot;</span>)
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(vcffile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(data.path, <span class="kw">basename</span>(vcffile)), vcffile)
gdsfile &lt;-<span class="st"> &quot;data/1KG_phase3_subset_chr1.gds&quot;</span>
<span class="kw">seqVCF2GDS</span>(vcffile, gdsfile, <span class="dt">fmt.import=</span><span class="st">&quot;GT&quot;</span>, <span class="dt">storage.option=</span><span class="st">&quot;LZMA_RA&quot;</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>We can interact with the GDS file using the SeqArray package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)
gds</code></pre></div>
<pre><code>## Object of class &quot;SeqVarGDSClass&quot;
## File: /projects/users/stephanie/Code/TOPMed/topmed_workshop_2018/data/1KG_phase3_subset_chr1.gds (70.6K)
## +    [  ] *
## |--+ description   [  ] *
## |--+ sample.id   { Str8 1126 LZMA_ra(9.66%), 877B } *
## |--+ variant.id   { Int32 1120 LZMA_ra(17.5%), 793B } *
## |--+ position   { Int32 1120 LZMA_ra(78.5%), 3.4K } *
## |--+ chromosome   { Str8 1120 LZMA_ra(4.55%), 109B } *
## |--+ allele   { Str8 1120 LZMA_ra(26.0%), 1.2K } *
## |--+ genotype   [  ] *
## |  |--+ data   { Bit2 2x1126x1121 LZMA_ra(8.34%), 51.4K } *
## |  |--+ extra.index   { Int32 3x0 LZMA_ra, 18B } *
## |  \--+ extra   { Int16 0 LZMA_ra, 18B }
## |--+ phase   [  ]
## |  |--+ data   { Bit1 1126x1120 LZMA_ra(0.11%), 177B } *
## |  |--+ extra.index   { Int32 3x0 LZMA_ra, 18B } *
## |  \--+ extra   { Bit1 0 LZMA_ra, 18B }
## |--+ annotation   [  ]
## |  |--+ id   { Str8 1120 LZMA_ra(40.4%), 3.6K } *
## |  |--+ qual   { Float32 1120 LZMA_ra(2.46%), 117B } *
## |  |--+ filter   { Int32,factor 1120 LZMA_ra(2.46%), 117B } *
## |  |--+ info   [  ]
## |  \--+ format   [  ]
## \--+ sample.annotation   [  ]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the unique sample identifier comes from the VCF header</span>
sample.id &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;sample.id&quot;</span>)
<span class="kw">length</span>(sample.id)</code></pre></div>
<pre><code>## [1] 1126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(sample.id)</code></pre></div>
<pre><code>## [1] &quot;HG00096&quot; &quot;HG00097&quot; &quot;HG00099&quot; &quot;HG00100&quot; &quot;HG00101&quot; &quot;HG00102&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># a unique integer ID is assigned to each variant</span>
variant.id &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;variant.id&quot;</span>)
<span class="kw">length</span>(variant.id)</code></pre></div>
<pre><code>## [1] 1120</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(variant.id)</code></pre></div>
<pre><code>## [1] 1 2 3 4 5 6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reference allele frequency of each variant</span>
afreq &lt;-<span class="st"> </span><span class="kw">seqAlleleFreq</span>(gds)
<span class="kw">hist</span>(afreq, <span class="dt">breaks=</span><span class="dv">50</span>)</code></pre></div>
<p><img src="_main_files/figure-html/seqarray-1.png" width="672" /></p>
<p>We can define a filter on the <code>gds</code> object. After using the <code>seqSetFilter</code> command, all subsequent reads from the <code>gds</code> object are restricted to the selected subset of data, until a new filter is defined or <code>seqResetFilter</code> is called.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.id=</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sample.id=</span>sample.id[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</code></pre></div>
<pre><code>## # of selected samples: 5
## # of selected variants: 10</code></pre>
<p>Genotype data is stored in a 3-dimensional array, where the first dimension is always 2 for diploid genotypes. The second and third dimensions are samples and variants, respectively. The values of the array denote alleles: <code>0</code> is the reference allele and <code>1</code> is the alternate allele. For multiallelic variants, other alternate alleles are represented as integers <code>&gt; 1</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;genotype&quot;</span>)
<span class="kw">dim</span>(geno)</code></pre></div>
<pre><code>## [1]  2  5 10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno[,,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</code></pre></div>
<pre><code>## , , 1
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 2
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0</code></pre>
<p>The <a href="http://bioconductor.org/packages/SeqVarTools">SeqVarTools package</a> has some additional functions for interacting with SeqArray-format GDS files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqVarTools)

<span class="co"># return genotypes in matrix format</span>
<span class="kw">getGenotype</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1     2     3     4     5     6     7     8     9     10   
##   HG00096 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00097 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00099 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00100 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00101 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getGenotypeAlleles</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1     2     3     4     5     6     7     8     9     10   
##   HG00096 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00097 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00099 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00100 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00101 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">refDosage</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1 2 3 4 5 6 7 8 9 10
##   HG00096 2 2 2 2 2 2 2 2 2  2
##   HG00097 2 2 2 2 2 2 2 2 2  2
##   HG00099 2 2 2 2 2 2 2 2 2  2
##   HG00100 2 2 2 2 2 2 2 2 2  2
##   HG00101 2 2 2 2 2 2 2 2 2  2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">altDosage</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1 2 3 4 5 6 7 8 9 10
##   HG00096 0 0 0 0 0 0 0 0 0  0
##   HG00097 0 0 0 0 0 0 0 0 0  0
##   HG00099 0 0 0 0 0 0 0 0 0  0
##   HG00100 0 0 0 0 0 0 0 0 0  0
##   HG00101 0 0 0 0 0 0 0 0 0  0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at reference and alternate alleles</span>
<span class="kw">refChar</span>(gds)</code></pre></div>
<pre><code>##  [1] &quot;C&quot; &quot;C&quot; &quot;C&quot; &quot;C&quot; &quot;T&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot; &quot;A&quot; &quot;C&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">altChar</span>(gds)</code></pre></div>
<pre><code>##  [1] &quot;G&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;T&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reset the filter to all variants and samples</span>
<span class="kw">seqResetFilter</span>(gds)</code></pre></div>
<pre><code>## # of selected samples: 1,126
## # of selected variants: 1,120</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># how many alleles for each variant?</span>
n &lt;-<span class="st"> </span><span class="kw">seqNumAllele</span>(gds)
<span class="kw">table</span>(n)</code></pre></div>
<pre><code>## n
##    2    3    4 
## 1099   20    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># some variants have more than one alternate allele</span>
multi.allelic &lt;-<span class="st"> </span><span class="kw">which</span>(n <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)
<span class="kw">altChar</span>(gds)[multi.allelic]</code></pre></div>
<pre><code>##  [1] &quot;GT,G&quot;            &quot;G,T&quot;             &quot;A,T&quot;            
##  [4] &quot;A,T&quot;             &quot;ATG,ATGTG&quot;       &quot;C,G&quot;            
##  [7] &quot;A,T&quot;             &quot;C,T&quot;             &quot;A,C&quot;            
## [10] &quot;TAA,T&quot;           &quot;GTTA,GTTT&quot;       &quot;GCC,GCCC,G&quot;     
## [13] &quot;A,C&quot;             &quot;A,C&quot;             &quot;A,C&quot;            
## [16] &quot;CAAGCAT,CGAGCAT&quot; &quot;CATTATT,C&quot;       &quot;AT,C&quot;           
## [19] &quot;TGTGA,C&quot;         &quot;CCATT,CCATTCATT&quot; &quot;C,G&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract a particular alternate allele</span>
<span class="co"># first alternate</span>
<span class="kw">altChar</span>(gds, <span class="dt">n=</span><span class="dv">1</span>)[multi.allelic]</code></pre></div>
<pre><code>##  [1] &quot;GT&quot;      &quot;G&quot;       &quot;A&quot;       &quot;A&quot;       &quot;ATG&quot;     &quot;C&quot;       &quot;A&quot;      
##  [8] &quot;C&quot;       &quot;A&quot;       &quot;TAA&quot;     &quot;GTTA&quot;    &quot;GCC&quot;     &quot;A&quot;       &quot;A&quot;      
## [15] &quot;A&quot;       &quot;CAAGCAT&quot; &quot;CATTATT&quot; &quot;AT&quot;      &quot;TGTGA&quot;   &quot;CCATT&quot;   &quot;C&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># second alternate</span>
<span class="kw">altChar</span>(gds, <span class="dt">n=</span><span class="dv">2</span>)[multi.allelic]</code></pre></div>
<pre><code>##  [1] &quot;G&quot;         &quot;T&quot;         &quot;T&quot;         &quot;T&quot;         &quot;ATGTG&quot;    
##  [6] &quot;G&quot;         &quot;T&quot;         &quot;T&quot;         &quot;C&quot;         &quot;T&quot;        
## [11] &quot;GTTT&quot;      &quot;GCCC&quot;      &quot;C&quot;         &quot;C&quot;         &quot;C&quot;        
## [16] &quot;CGAGCAT&quot;   &quot;C&quot;         &quot;C&quot;         &quot;C&quot;         &quot;CCATTCATT&quot;
## [21] &quot;G&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># how many variants are SNVs vs INDELs?</span>
<span class="kw">table</span>(<span class="kw">isSNV</span>(gds, <span class="dt">biallelic=</span><span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   110  1010</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">isSNV</span>(gds, <span class="dt">biallelic=</span><span class="ot">FALSE</span>))</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##    99  1021</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 11 SNVs are multi-allelic</span></code></pre></div>
</div>
<div id="exercises" class="section level2">
<h2><span class="header-section-number">2.2</span> Exercises</h2>
<ol style="list-style-type: decimal">
<li><p>Set a filter selecting only multi-allelic variants. Inspect their genotypes using the different methods you learned above. Use the <code>alleleDosage</code> method to find dosage for the second (and third, etc.) alternate allele.</p></li>
<li><p>Use the <code>hwe</code> function in SeqVarTools to run a Hardy-Weinberg Equilibrium test on each variant. Identify a variant with low p-value and inspect its genotypes. (Note that the HWE test is only valid for biallelic variants, and will return <code>NA</code> for multiallelic variants.)</p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="computing-a-grm.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
